src/kaleidoscope/plugin/LED-ActiveModColor.cpp

```diff
@@ -32,57 +32,67 @@ cRGB ActiveModColorEffect::highlight_color = (cRGB) {
 
 cRGB ActiveModColorEffect::sticky_color = CRGB(160, 0, 0);
 
-EventHandlerResult ActiveModColorEffect::onLayerChange() {
-  if (!Runtime.has_leds)
+EventHandlerResult ActiveModColorEffect::onKeyswitchEvent(
+  Key &key,
+  KeyAddr key_addr,
+  uint8_t key_state) {
+  // If `key_addr` is not a physical key address, ignore it:
+  if (! key_addr.isValid()) {
     return EventHandlerResult::OK;
+  }
 
-  mod_key_count_ = 0;
-
-  for (auto key_addr : KeyAddr::all()) {
-    Key k = Layer.lookupOnActiveLayer(key_addr);
-
-    if (::OneShot.isOneShotKey(k) ||
-        (highlight_normal_modifiers_ && (
-           (k >= Key_LeftControl && k <= Key_RightGui) ||
-           (k.getFlags() == (SYNTHETIC | SWITCH_TO_KEYMAP))))) {
-      mod_keys_[mod_key_count_++] = key_addr;
```

Comment:
This is where it's possible for the existing version to write past the end of the `mod_keys_` array, due to the lack of bounds checking. It's unlikely that a keymap would contain more than 16 modifiers, but if we extended the current design to cover DualUse Qukeys, then it would be quite plausible to have a keymap that goes over this limit. With a full complement of standard (or OneShot) modifiers, plus a full set of DualUse Qukey modifiers, we're already at 16. Add to that any layer shift keys, and it wouldn't be difficult to go over the limit.